File: web/src/aggregation.ts
/** Sum official census-unit counts; weight only units cut by the selection border. */
export interface SelectedUnit {
  key: string
  counts: Record<string, number>
  coverage?: number
}

export interface Aggregate {
  counts: Record<string, number>
  quality: 'exacto' | 'estimado'
  estimated_percent: number
}

export function aggregate(units: SelectedUnit[]): Aggregate {
  const values: Record<string, number> = {}
  let estimatedPopulation = 0
  let partial = false
  for (const unit of units) {
    const coverage = unit.coverage ?? 1
    if (!Number.isFinite(coverage) || coverage < 0 || coverage > 1) {
      throw new Error(`Invalid area coverage for ${unit.key}`)
    }
    for (const [name, count] of Object.entries(unit.counts)) {
      if (!Number.isFinite(count) || count < 0) {
        throw new Error(`Negative or invalid census count for ${unit.key}.${name}`)
      }
      values[name] = (values[name] ?? 0) + count * coverage
    }
    if (coverage > 0 && coverage < 1) {
      partial = true
      estimatedPopulation += (unit.counts.population ?? 0) * coverage
    }
  }
  const population = values.population ?? 0
  return {
    counts: values,
    quality: partial ? 'estimado' : 'exacto',
    estimated_percent: population > 0 ? 100 * estimatedPopulation / population : 0,
  }
}

File: pipeline/aggregation.py
"""Add published census-unit counts, weighting only boundary-cut manzanas."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class SelectedUnit:
    key: str
    counts: dict[str, float]
    coverage: float = 1.0


@dataclass(frozen=True)
class Aggregate:
    counts: dict[str, float]
    quality: str
    estimated_percent: float


def aggregate(units: list[SelectedUnit]) -> Aggregate:
    """Sum complete units exactly; use area share for units cut by a boundary."""
    values: dict[str, float] = {}
    estimated_population = 0.0
    for unit in units:
        if not 0 <= unit.coverage <= 1:
            raise ValueError(f"Invalid area coverage for {unit.key}")
        for name, count in unit.counts.items():
            if count < 0:
                raise ValueError(f"Negative census count for {unit.key}.{name}")
            values[name] = values.get(name, 0.0) + count * unit.coverage
        if 0 < unit.coverage < 1:
            estimated_population += unit.counts.get("population", 0.0) * unit.coverage
    population = values.get("population", 0.0)
    partial = any(0 < unit.coverage < 1 for unit in units)
    return Aggregate(
        values,
        "estimado" if partial else "exacto",
        100 * estimated_population / population if population > 0 else 0.0,
    )
